prov/terraform/openstack/generator.go

```diff
@@ -103,146 +108,250 @@ func (g *osGenerator) GenerateTerraformInfraForNode(ctx context.Context, cfg con
+	default:
+		err = errors.Errorf("Unsupported node type '%s' for node '%s' in deployment '%s'",
+			opts.nodeType, opts.nodeName, opts.deploymentID)
+	}
+	return err
+}
 
-		case "yorc.nodes.openstack.Network":
-			networkID, err := deployments.GetNodePropertyValue(kv, deploymentID, nodeName, "network_id")
-			if err != nil {
-				return false, nil, nil, nil, err
-			} else if networkID != nil && networkID.RawString() != "" {
-				log.Debugf("Reusing existing volume with id %q for node %q", networkID, nodeName)
-				return false, nil, cmdEnv, nil, nil
-			}
-			var network Network
-			network, err = g.generateNetwork(kv, cfg, deploymentID, nodeName)
+func (g *osGenerator) generateBlockStorageInfra(opts generateInfraOptions) error {
 
-			if err != nil {
-				return false, nil, nil, nil, err
-			}
-			var subnet Subnet
-			subnet, err = g.generateSubnet(kv, cfg, deploymentID, nodeName)
+	var bsIds []string
+	volumeID, err := deployments.GetNodePropertyValue(opts.kv, opts.deploymentID, opts.nodeName, "volume_id")
+	if err != nil {
+		return err
+	}
+
+	if volumeID != nil && volumeID.RawString() != "" {
+		log.Debugf("Reusing existing volume with id %q for node %q", volumeID, opts.nodeName)
+		bsIds = strings.Split(volumeID.RawString(), ",")
+	}
+
+	var bsVolume BlockStorageVolume
+	bsVolume, err = g.generateOSBSVolume(opts.kv, opts.cfg, opts.deploymentID, opts.nodeName, opts.instanceName)
+	if err != nil {
+		return err
+	}
 
+	if len(bsIds)-1 < opts.instanceIndex {
+		commons.AddResource(opts.infrastructure, opts.resourceTypes[blockStorageVolume], bsVolume.Name, &bsVolume)
+		consulKey := commons.ConsulKey{
+			Path:  path.Join(opts.instancesKey, opts.instanceName, "/attributes/volume_id"),
+			Value: fmt.Sprintf("${%s.%s.id}", opts.resourceTypes[blockStorageVolume], bsVolume.Name)}
+		consulKeys := commons.ConsulKeys{Keys: []commons.ConsulKey{consulKey}}
+		commons.AddResource(opts.infrastructure, consulKeysResource, bsVolume.Name, &consulKeys)
+	} else {
+		name := opts.cfg.ResourcesPrefix + opts.nodeName + "-" + opts.instanceName
+		consulKey := commons.ConsulKey{
+			Path:  path.Join(opts.instancesKey, opts.instanceName, "/properties/volume_id"),
+			Value: strings.TrimSpace(bsIds[opts.instanceIndex])}
+		consulKeys := commons.ConsulKeys{Keys: []commons.ConsulKey{consulKey}}
+		commons.AddResource(opts.infrastructure, consulKeysResource, name, &consulKeys)
+	}
+
+	return err
+}
+
+func (g *osGenerator) generateFloatingIPInfra(opts generateInfraOptions) error {
+
+	ip, err := g.generateFloatingIP(opts.kv, opts.deploymentID, opts.nodeName, opts.instanceName)
+	if err != nil {
+		return err
+	}
+
+	var consulKey commons.ConsulKey
+	if !ip.IsIP {
+		floatingIP := FloatingIP{Pool: ip.Pool}
+		commons.AddResource(opts.infrastructure, opts.resourceTypes[computeFloatingIP], ip.Name, &floatingIP)
+		consulKey = commons.ConsulKey{
+			Path:  path.Join(opts.instancesKey, opts.instanceName, floatingIPEndpointCapAttribute),
```

Comment:
Not sure but by doing this way, will we have attributes notifications works correctly?
Dealing with outputs and publishing attributes changes at the end of the process wouldn't be safer?
WDYT @stebenoist?

Reply:
Nothing seems to have changed here

Reply:
yes, just moved this code to a function, to fix a Sonar issue about the complexity of the main function